File: DataStructures/LinearDataStructures/C++/stack.cpp
namespace data_structures {
#include<iostream>

template<class T>
class Stack {
public:
    Stack();
    ~Stack();

    void push(T);
    T pop();
    T peek();

    bool isEmpty();

    unsigned int getSize();

private:
    LinkedList<T> *list;
};


template<class T> Stack<T>::Stack() {
    this->list = new LinkedList<T>();
}

template<class T> Stack<T>::~Stack() {
    delete this->list;
}

template<class T>
void Stack<T>::push(T value) {
    this->list->addTail(value);
}

template<class T>
T Stack<T>::pop() {
    T value = this->list->getTail();
    this->list->removeTail();
    return value;
}

template<class T>
T Stack<T>::peek() {
    return this->list->getTail();
}

template<class T>
unsigned  int  Stack<T>::getSize() {
    return this->list->getSize();
}

template<class T>
bool Stack<T>::isEmpty() {
    return this->getSize() == 0;
}

}

File: DataStructures/LinearDataStructures/Stack.cs
namespace LinearDataStructures
{
    public class Stack<T>
    {
        private readonly LinkedList<T> list;

        public Stack()
        {
            this.list = new LinkedList<T>();
        }

        public int Count
        {
            get
            {
                return this.list.Count;
            }
        }

        public void Push(T value)
        {
            this.list.AddLast(value);
        }

        public T Pop()
        {
            var value = this.list.Last;
            this.list.RemoveLast();
            return value;
        }

        public T Peek()
        {
            return this.list.Last;
        }

        public bool IsEmpty()
        {
            return this.Count == 0;
        }
    }
}
